File: src/main/java/com/devs/roamance/model/travel/journal/Journal.java
package com.devs.roamance.model.travel.journal;

import com.devs.roamance.model.audit.Audit;
import com.devs.roamance.model.common.Location;
import com.devs.roamance.model.user.User;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import jakarta.annotation.Nullable;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@Setter
@NoArgsConstructor
@RequiredArgsConstructor
@AllArgsConstructor
@Entity
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@EntityListeners(AuditingEntityListener.class)
@Table(
    name = "journals",
    uniqueConstraints = @UniqueConstraint(columnNames = {"title", "created_by"}))
public class Journal {

  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;

  @NonNull private String title;

  @NonNull @Embedded private Location destination;

  @Nullable
  @Column(length = 1000)
  private String description;

  private Boolean isFavorite;
  private Boolean isArchived;
  private Boolean isShared;
  private LocalDate date;

  @Nullable private String coverImage;

  @OneToMany(
      mappedBy = "journal",
      fetch = FetchType.LAZY,
      cascade = CascadeType.ALL,
      orphanRemoval = true)
  private List<Subsection> subsections = new ArrayList<>();

  @JsonIgnore
  @ManyToOne(
      fetch = FetchType.LAZY,
      cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.DETACH, CascadeType.REFRESH})
  @JoinColumn(name = "user_id", referencedColumnName = "id")
  private User user;

  @Embedded private Audit audit = new Audit();

  public void addSubsection(Subsection subsection) {
    subsections.add(subsection);
    subsection.setJournal(this);
  }

  public void removeSubsection(Subsection subsection) {
    subsections.remove(subsection);
  }
}

File: roamance/src/types/journal.ts
import { Audit } from './audit';
import { Location } from './location';
import { BaseResponse } from './response';
import {
  SubsectionRequest,
  SightseeingSubsectionDto,
  ActivitySubsectionDto,
  RouteSubsectionDto,
  Subsection,
} from './subsection';
import { User } from './user';

export interface Journal {
  id: string;
  title: string;
  destination: Location;
  description: string;
  subsections: Subsection[];
  user: User;
  audit: Audit;
}

/* --------------------------------- Create --------------------------------- */

export interface JournalCreateRequest {
  title: string;
  destination: Location;
  description: string;
  subsections: SubsectionRequest[];
}

/* -------------------------------- Response -------------------------------- */

export interface JournalDto {
  id: string;
  title: string;
  destination: Location;
  description: string;
  is_favorite: boolean;
  is_archived: boolean;
  is_shared: boolean;
  date: string;
  cover_image: string;
  audit: Audit;
}

export interface JournalBrief extends JournalDto {
  total_subsections: number;
}

export interface JournalDetail extends JournalDto {
  subsections: (
    | SightseeingSubsectionDto
    | ActivitySubsectionDto
    | RouteSubsectionDto
  )[];
}

export type JournalBriefResponse = BaseResponse<JournalBrief>;
export type JournalListResponse = BaseResponse<JournalBrief[]>;
export type JournalDetailResponse = BaseResponse<JournalDetail>;
